### ISSUE TITLE
Pretrained VAE?

### ISSUE THREAD

**Issue by AliBaheri** [r7vme/learning-to-drive-in-a-day#2]
Hi,
Great work.
A quick question- How did you train VAE? I mean which data-set has been used to train VAE?

**Comment by r7vme** [comment]
Hi @AliBaheri 

> How did you train VAE? I mean which data-set has been used to train VAE?

To get dataset (images from simulator), just save observations (numpy array 80x160x3) [here](https://github.com/r7vme/learning-to-drive-in-a-day/blob/9e001d83405bbbe5aa777e3653320d55bd8814ac/donkey_gym_wrapper/env.py#L86). If you specifically need dataset.

At the moment it's not necessary as you can just run a full training as it is and at the end it produce `vae.json`. Under the hood it saves observations (images) to the VAE [controller buffer](https://github.com/r7vme/learning-to-drive-in-a-day/blob/9e001d83405bbbe5aa777e3653320d55bd8814ac/donkey_gym_wrapper/env.py#L91) and then [training run online](https://github.com/r7vme/learning-to-drive-in-a-day/blob/9e001d83405bbbe5aa777e3653320d55bd8814ac/vae/controller.py#L91-L114) after [every episode](https://github.com/r7vme/learning-to-drive-in-a-day/blob/9e001d83405bbbe5aa777e3653320d55bd8814ac/ddpg_with_vae.py#L92).

**Comment by AliBaheri** [comment]
Great, thanks for your response.

Would it be possible to give some information about reward function? I see in your original post you said a special reward function. What do you mean by that? BTW, what do you call it learn in **50 sec** (or 20 mins or so) or WAYVE call it drive in **a day** regarding this fact that A2C / A3C suffer from sample efficiency?

In fact, WAYVE defines the reward as:

**Distance traveled by the car before  a safety driver takeover against number of exploration episodes.**

How the reward is defined in your environment?

Thanks.
Best regards,

**Comment by r7vme** [comment]
> Would it be possible to give some information about reward function? I see in your original post you said a special reward function. What do you mean by that? BTW, what do you call it learn in 50 sec (or 20 mins or so) or WAYVE call it drive in a day regarding this fact that A2C / A3C suffer from sample efficiency?

Reward function defined in donkey_gym repo [here](https://github.com/r7vme/donkey_gym/blob/377343a840b462dfd54f6abb62eeeb58078976c2/donkey_gym/envs/donkey_sim.py#L183-L188). I'm using car "velocity" or to be more precise distance traveled in this discrete time period (one step). Terminal step has reward zero.

Regarding 50 sec / 20 minutes. 50 sec was with pretrained VAE, so in 50 seconds (~3-5 episodes) DDPG learned decent policy. 20 minutes it's roughly time that current script шs spending to train both VAE and DDPG simultaneously.


> In fact, WAYVE defines the reward as: Distance traveled by the car before a safety driver takeover against number of exploration episodes.

This is a question i also asked them in email (unfortunately they can not disclose any info).  In paper reward defined like this

> Instead, we define the reward as forward speed and terminate an episode upon an infraction of traffic rules – thus the value of a given state V (st) corresponds to the average distance travelled before an infraction.

reward - forward speed, OK. And then their interpretation that this will correspond to distance travelled by car before infracion. So now how i understood this.

Look on equation (1) in the paper. Final value for state s_0 (or e.g. s_t) is sum of all rewards afterwards. Equation (1) sums by "infinity", but we need a finite summarization, because we have terminal state. This what they describe in next paragraph.

> In our setting, we use a finite time horizon T in place of infinity in the above formula. This is equivalent to one of the states being terminal, i.e. it cannot be escaped and any action at that state gives zero reward.

So to summarize:
reward - forward speed
final value for the state - distance travelled from this state till infraction.

**Comment by r7vme** [comment]
> this fact that A2C / A3C suffer from sample efficiency?

need to read up about A2C / A3C

**Comment by AliBaheri** [comment]
Thanks for clarification.

**Comment by AliBaheri** [comment]
A quick question - I just noticed one thing in Read Me file - Are you optimize VAE after each episode?

**Comment by r7vme** [comment]
> A quick question - I just noticed one thing in Read me file - Are you optimize VAE after each episode?

Yes, after every episode and it's [here](https://github.com/r7vme/learning-to-drive-in-a-day/blob/master/ddpg_with_vae.py#L94).
